File: properties.py
from enum import Enum
import sys
from jsonfile import JSONFile
from pprint import pprint
import collections
from typing import Optional, Union, Any, List, Dict
import json
import abc

class PropNodeType(Enum):
    group = 0
    unsigned_int = 1
    int = 2
    unsigned_float = 3
    bool = 4
    rgb = 5
    hsv = 6
    string = 7
    float = 8


class PropertyNode:
    PATH_SEPARATOR = '/'

    def __init__(self, name, type, parent: "PropertyNode"=None):
        self.type = type
        self.name = name  # todo - validate self.name for invalid paths
        self.parent = parent

    def as_description(self) -> collections.OrderedDict:
        return collections.OrderedDict(
            [("name", self.name),
             ("type", PropNodeType(self.type).name.lower())
             ])

    def as_dict(self) -> Dict:
        return {
            self.name: self.contents()
        }

    def contents(self):
        return None

    def path(self):
        p = [self.name]
        node = self.parent
        while node is not None:
            if node.parent is not None:
                p.append(node.name)
                node = node.parent
            else:
                node = None
        p.reverse()
        return p

    def path_as_str(self) -> str:
        return PropertyNode.PATH_SEPARATOR.join(self.path())

    def root(self):
        seek = self
        while seek is not None:
            if seek.parent is None:
                return seek
            else:
                seek = seek.parent

    def on_changed(self, prop, is_dirty=False, from_runtime_change=True):
        pass


class Property(PropertyNode):
    numerics = [PropNodeType.unsigned_int,
                PropNodeType.int,
                PropNodeType.float,
                PropNodeType.unsigned_float]

    numeric_tuple = [PropNodeType.rgb, PropNodeType.hsv]

    def __init__(self, name, type, parent=None, default=None, value=None):
        super().__init__(name, type, parent)
        self.default = default
        self.min = None
        self.max = None
        self._value = None
        self.names = []
        self.dirty = False

        if self.type == PropNodeType.unsigned_int:
            self.min = 0
            self.max = sys.maxsize
        elif self.type == PropNodeType.int:
            self.min = -sys.maxsize
            self.max = sys.maxsize
        elif self.type == PropNodeType.float:
            self.min = float("-inf")
            self.max = float("inf")
        elif self.type == PropNodeType.unsigned_float:
            self.min = 0.0
            self.max = float("inf")
        elif self.type == PropNodeType.bool:
            self.min = 0
            self.max = 1
        elif self.type == PropNodeType.rgb:
            self.min = (0, 0, 0)
            self.max = (255, 255, 255)
            self.names = ["red", "green", "blue"]
        elif self.type == PropNodeType.hsv:
            self.min = (0.0, 0.0, 0.0)
            self.max = (360.0, 1.0, 1.0)
            self.names = ["hue", "saturation", "luminosity"]

        if self.default is None:
            if self.is_single_numeric():
                self.default = 0
            if self.type == PropNodeType.bool:
                self.default = False
            if self.type in self.numeric_tuple:
                self.default = (0, 0, 0)

        if value is not None:
            self.set(value, from_runtime_change=False)
        else:
            self.set(self.default, from_runtime_change=False)

    def as_description(self):
        d = super().as_description()
        d["min"] = self.min
        d["max"] = self.max
        d["default"] = self.default
        d["names"] = self.names
        return d

    def range(self, minimum=None, maximum=None):
        if minimum is not None:
            self.min = minimum
        if maximum is not None:
            self.max = maximum
        return self.min, self.max

    def is_single_numeric(self):
        return self.type in Property.numerics

    @staticmethod
    def clip(val, mi, ma):
        if val < mi:
            return mi
        if val > ma:
            return ma
        return val

    # returns True if value was changed
    def set(self, value, index=None, from_runtime_change: bool=True):
        temp_value = self.value()
        if self.is_single_numeric():
            self._value = self.clip(value, self.min, self.max)
        if self.type in Property.numeric_tuple:
            if (index is not None) and (index < len(self.min)):
                t = list(self._value)
                t[index] = self.clip(value, self.min[index], self.max[index])
                self._value = tuple(t)
            elif len(value) == len(self.min):
                p = []
                # surely there is a more python way...
                for v, mi, ma in zip(value, self.min, self.max):
                    p.append(self.clip(v, mi, ma))
                self._value = tuple(p)

        if self.type == PropNodeType.bool:
            if value == 0 or value is False:
                self._value = False
            elif value == 1 or value is True:
                self._value = True
            else:
                self._value = self.default

        was_changed = temp_value != self.value()
        if from_runtime_change:
            self.dirty = was_changed or self.dirty

        if was_changed:
            self.root().on_changed(self, self.dirty, from_runtime_change)

        return was_changed

    def value(self):
        return self._value

    def clean(self):
        self.dirty = False

    def is_dirty(self):
        return self.dirty

    def from_contents(self, contents):
        if (self.is_single_numeric() or
                (self.type == PropNodeType.bool) or
                (self.type == PropNodeType.string)):
            self.set(contents, from_runtime_change=False)
        elif self.type in Property.numeric_tuple:
            self.set(tuple(contents), from_runtime_change=False)

    def contents(self):
        if (self.is_single_numeric() or
                (self.type == PropNodeType.bool) or
                (self.type == PropNodeType.string)):
            return self.value()
        elif self.type in Property.numeric_tuple:
            return list(self.value())
        pass


class PropertiesListener(metaclass=abc.ABCMeta):

    @abc.abstractmethod
    def on_prop_updated(self, prop: PropertyNode, from_runtime_change: bool=True):
        pass


class Properties(PropertyNode):
    def __init__(self, name: str, parent=None, window_name=None):
        super().__init__(name, PropNodeType.group, parent)
        self.children = collections.OrderedDict()  # Ordered because adding is hierarchical
        self.file_path = None
        self.auto_file_save = True
        self.window_name = self.name
        self.on_changed_handler = None
        self.on_changed_listeners = []

    def child_nodes(self) -> List[Union[PropertyNode, Property, "Properties"]]:
        return [x for key, x in self.children.items()]

    def add_group(self, name: str) -> "Properties":
        """Add a group node that will contain subnodes, but is itself contains no data

        :param name:
        :return: Property
        """
        return self.add(name, PropNodeType.group)

    def add(self, name: str, prop_type: PropNodeType, default=None, minimum=None, maximum=None):
        if prop_type is PropNodeType.group:
            p = Properties(name, self)
        else:
            p = Property(name, prop_type, self, default, default)
            p.range(minimum, maximum)

        self.children[name] = p
        return p

    def add_listener(self, listener: PropertiesListener):
        self.on_changed_listeners.append(listener)


    def contents(self) -> collections.OrderedDict:
        """Returns the contents, i.e. structure AND data of this node and subnodes.
        It does NOT return a named-root. i.e. the calling nodes name is not included. use
        from_dict() instead.

        :return: OrderedDict
        """
        d = collections.OrderedDict()
        for node in self.child_nodes():
            d[node.name] = node.contents()
        return d

    def contents_json(self):
        return json.dumps(self.contents(), indent=2)

    def from_contents(self, contents: collections.OrderedDict):
        for node in self.child_nodes():
            if node.name in contents:
                node.from_contents(contents[node.name])

    def from_contents_json(self, contents_json):
        self.from_contents(json.loads(contents_json))

    def from_dict(self, dictionary: dict):
        self.from_contents(dictionary[self.name])

    def save(self, path, file_type="json"):
        if file_type == "json":
            d = self.contents()
            JSONFile.save(path, d)
        pass

    def is_dirty(self) -> bool:
        for node in self.child_nodes():
            if node.is_dirty():
                return True
        return False

    def clean(self):
        for node in self.child_nodes():
            node.clean()

    def update_permanent_storage(self):
        if self.file_path is not None and self.is_dirty():
            self.save(self.file_path)
            self.clean()

    def load(self, path: str, file_type="json"):
        if file_type == "json":
            data = JSONFile.load(path)
            self.file_path = path
            if data is not None:
                self.from_contents(data)
                print("loaded: " + self.file_path)
            else:
                self.from_contents(self.contents())
                print("loaded from defaults")
        pass

    def node_path_list_of(self, node_path: Union[List[str], str]) -> List[str]:
        """From varied input return a list of nodenames in order of hierarchy.

        """
        if isinstance(node_path, str):
            return node_path.split('/')
        return node_path

    def node_at_path(self, node_path: Union[List[str], str]) -> Optional[PropertyNode]:
        node_path_list = self.node_path_list_of(node_path)
        current_node = self
        for node_name in node_path_list:
            current_node = current_node.named_node(node_name)
            if current_node is None:
                print("Node not found: " + "/".join(node_path_list))
                return None
            if current_node.type != PropNodeType.group:  # actual data node (i.e. LEAF) so return as CANNOT continue.
                return current_node
            else:
                pass  # i.e. go one deeper in hierarchy.
        return current_node

    def named_node(self, node_name: str) -> Optional[PropertyNode]:
        if node_name in self.children:
            return self.children[node_name]
        return None

    def property_at_path(self, node_path) -> Optional[Property]:
        node = self.node_at_path(node_path)
        if node is not None:
            if isinstance(node, Property):
                return node
            else:
                print("Valid nodepath but a group node not a property")
        return None

    def group_at_path(self, node_path) -> Optional["Properties"]:
        # https://blog.jetbrains.com/pycharm/2015/11/python-3-5-type-hinting-in-pycharm-5/
        node = self.node_at_path(node_path)
        if node is not None:
            if isinstance(node, Properties):
                return node
            else:
                print("Valid nodepath but a property not a group")
        return None

    def value_of(self, node_path: Union[List[str], str]) -> Any:
        node = self.property_at_path(node_path)
        if node is not None:
            return node.value()

    def set_value_of(self, node_path: Union[List[str], str], value: Any, from_run_time: bool = True) -> Any:
        node = self.property_at_path(node_path)
        if node is not None:
            return node.set(value=value, from_runtime_change=from_run_time)

    def as_description(self) -> collections.OrderedDict:
        """Returns a complete description of all propeties and hierarchy but NO DATA.

        :return:
        """
        prop_descriptions = [p.as_description() for key, p in self.children.items()]
        d = super().as_description()
        d["children"] = prop_descriptions
        return d

    def as_description_json(self):
        return json.dumps(self.as_description(), indent=2)

    def on_changed(self, prop, is_dirty=False, from_runtime_change=True):
        # print("changed "+prop.name)
        if is_dirty and self.auto_file_save is True:
            self.update_permanent_storage()  # todo don't auto save like, this but periodically check to reduce writes
        if self.on_changed_handler is not None:
            self.fire(self.on_changed_handler, prop, from_runtime_change)
        for listener in self.on_changed_listeners:
            listener.on_prop_updated(prop, from_runtime_change)

    def fire(self, handler, *args):
        handler(*args)


if __name__ == "__main__":
    props = Properties('test')
    props.on_changed_handler = lambda prop, x: print("changed: "+prop.name+' to: '+str(prop.value()))

    ui = props.add_group("ui")
    ui.add("blur", PropNodeType.unsigned_int, maximum=250)
    ui.add("min_circle", PropNodeType.unsigned_int, maximum=250)
    ui.add("max_circle", PropNodeType.unsigned_int, maximum=250, default=10)
    ui.add("show_masks", PropNodeType.bool, default=True)
    ui.add("min_area", PropNodeType.unsigned_int)
    colours = props.add_group("colours")
    col = ["blue"]
    for name in col:
        colour = colours.add_group(name)
        colour.add('min_hsv', PropNodeType.hsv)
        colour.add('max_hsv', PropNodeType.hsv, (180.0, 0.5, 0.5))

    print("DESCRIPTION as JSON as saved from file (should be same)\n-------------")
    print(props.as_description_json())
    print("Contents as PYTHON\n-------------")
    pprint(props.contents())
    print("Contents as JSON\n-------------")
    json_contents = props.contents_json()
    print(json_contents)
    props.from_contents_json(json_contents)
    print("Contents as JSON as loaded from Above (should be same)\n-------------")
    print(props.contents_json())
    props.save('test.json')
    props.load('test.json')
    print("Contents as JSON as saved from file (should be same)\n-------------")
    print(props.contents_json())
    print(props.value_of("ui/max_circle"))
    props.set_value_of('ui/max_circle', 50)
    print(props.value_of("ui/max_circle"))
    props.set_value_of('ui/max_circle', 450)
    print(props.value_of("ui/max_circle"))

    # print(json.dumps(props.as_dict(), indent=2))

File: properties.js
const PropNodeType = {
    group: 0,
    unsigned_int: 1,
    int: 2,
    unsigned_float: 3,
    bool: 4,
    rgb: 5,
    hsv: 6,
    string: 7,
    float: 8
};

const PropTypeSingleNumeric = [
    PropNodeType.unsigned_int,
    PropNodeType.int,
    PropNodeType.float,
    PropNodeType.unsigned_float
];

const PropTypeMultipleNumeric = [
    PropNodeType.rgb,
    PropNodeType.hsv,
];

class PropNode {
    constructor(name, type, parent = null) {
        this.name = name;
        this.type = type;
        /**
         * @var PropNode
         */
        this.parent = parent;
    }
    
    contents() {
        return null;
    }
    
    as_description() {
        let description = new Map();
        description.set("name", this.name);
        description.set("type", this.key_from_value(PropNodeType, this.type));
        return description;
    }
    
    key_from_value(object, value) {
        // https://stackoverflow.com/questions/36448649/get-key-corresponding-to-value-in-js-object
        return Object.keys(object).find(key => object[key] === value);
    }
    
    as_dict() {
        let d = {};
        d[this.name] = this.contents()
    }
    
    path() {
        let p = [this.name];
        let node = this.parent;
        while (node !== null) {
            if (node.parent !== null) {
                p.push(node.name);
                node = node.parent;
            } else {
                node = null;
            }
        }
        return p.reverse();
    }
    
    
}

class Property extends PropNode {
    constructor(name, type, parent=null, _default = undefined, value = undefined) {
        super(name, type, parent);
        this.default = _default;
        this.min = undefined;
        this.max = undefined;
        this._value = undefined;
        this.names = [];
        this.dirty = false;
        
        switch (type) {
            case PropNodeType.unsigned_int:
                this.min = 0;
                this.max = Number.MAX_SAFE_INTEGER;
                break;
            case PropNodeType.int:
                this.min = Number.MIN_SAFE_INTEGER;
                this.max = Number.MAX_SAFE_INTEGER;
                break;
            case PropNodeType.float:
                this.min = Number.NEGATIVE_INFINITY;
                this.max = Number.POSITIVE_INFINITY;
                break;
            case PropNodeType.unsigned_float:
                this.min = 0.0;
                this.max = Number.POSITIVE_INFINITY;
                break;
            case PropNodeType.bool:
                this.min = 0;
                this.max = 1;
                break;
            case PropNodeType.rgb:
                this.min = [0, 0, 0];
                this.max = [255, 255, 255];
                this.names = ["red", "green", "blue"];
                break;
            case PropNodeType.hsv:
                this.min = [0, 0, 0];
                this.max = [180, 255, 255];
                this.names = ["hue", "saturation", "luminosity"];
                break;
            default:
                console.log("Unknown PropType: type")
        }
        if (this.default === undefined) {
            if (this.is_single_numeric()) {
                this.default = 0;
            } else if (this.type === PropNodeType.bool) {
                this.default = false;
            } else if (this.is_multiple_numeric()) {
                this.default = [0, 0, 0]; //todo: convert to float if float.?
            }
        }
        if (value !== undefined) {
            this.set(value);
        } else {
            this.set(this.default);
        }
        
    }
    
    is_single_numeric() {
        return PropTypeSingleNumeric.includes(this.type); //
    }
    
    is_multiple_numeric() {
        return PropTypeMultipleNumeric.includes(this.type);
    }
    
    range(minimum = undefined, maximum = undefined) {
        if (minimum !== undefined) {
            this.min = minimum;
        }
        if (maximum !== undefined) {
            this.max = maximum;
        }
        return [this.min, this.max];
    }
    
    static clip(val, mi, ma) {
        if (val < mi) {
            return mi;
        } else if (val > ma) {
            return ma;
        }
        return val;
    }
    
    set(value, index = undefined, from_run_time_change = false) {
        // TODO: check TYPES
        let temp_value = this.value();
        if (this.is_single_numeric()) {
            this._value = Property.clip(value, this.min, this.max);
        }
        else if (this.is_multiple_numeric()) {
            if ((index !== undefined) && (index < this.min.length)) {
                this._value[index] = Property.clip(value);
            } else if (value.length === this.min.length) {
                this._value = value.map((val, index) => {
                    return Property.clip(val, this.min[index], this.max[index]);
                });
            }
        }
        else if (this.type === PropNodeType.bool) {
            if ((value === true) || (value === 1)) {
                this._value = true;
            } else if ((value === false) || (value === 0)) {
                this._value = false;
            } else {
                console.log("bool not set coz not valid bool value: " + this.name);
            }
        }
        
        let was_changed = false;
        if (this.is_multiple_numeric()) {
            was_changed = !Property.are_numeric_arrays_the_same(temp_value, this._value);
        } else {
            was_changed = (temp_value !== this._value)
        }
        if (from_run_time_change) {
            this.dirty = was_changed || this.dirty;
        }
        return was_changed;
    }
    
    value() {
        return this._value;
    }
    
    static are_numeric_arrays_the_same(ar1, ar2) {
        if ((ar1 === undefined) || (ar1 === undefined)) {
            return false;
        }
        if (ar1.length !== ar2.length) {
            return false;
        }
        for (let index = 0; index < ar1.length; index++) {
            if (ar1[index] !== ar2[index]) {
                return false;
            }
        }
        return true;
    }
    
    as_description() {
        let d = super.as_description();
        d["min"] = this.min;
        d["max"] = this.max;
        d["default"] = this.default;
        d["names"] = this.names;
        return d;
    }
    
    clean() {
        this.dirty = false;
    }
    
    is_dirty() {
        return this.dirty;
    }
    
    from_contents(contents) {
        this.set(contents)
    }
    
    contents() {
        return this.value();
    }
    
    
}

class Properties extends PropNode {
    constructor(name, parent=null) {
        super(name, PropNodeType.group, parent);
        this.children = new Map();
    }
    
    child_nodes() {
        return this.children; //does MAP support iterate?
    }
    
    add_group(name) {
        return this.add(name, PropNodeType.group);
    }
    
    add(name, type, _default = undefined, minimum = undefined, maximum = undefined) {
        let p = null;
        if (type === PropNodeType.group) {
            p = new Properties(name, this)
        } else {
            p = new Property(name, type, this, _default);
            p.range(minimum, maximum);
        }
        this.children.set(name, p);
        return p;
    }
    
    contents() {
        let d = new Map();
        this.child_nodes().forEach(node => {
            d.set(node.name, node.contents());
        });
        return d;
    }
    
    contents_json() {
        return value_to_json(this.contents());
    }
    
    from_contents(contents) {
        this.child_nodes().forEach(node => {
            if (node.name in contents) { //Might need to be more robust. but should be POJO
                node.from_contents(contents[node.name]);
            }
        })
    }
    
    from_contents_json(contents_json) {
        let contents = JSON.parse(contents_json);
        this.from_contents(contents);
    }
    
    from_dict(dict) {
        this.from_contents(dict[this.name]);
    }
    
    save() {
        console.log('not implementd');
    }
    
    load() {
        console.log('not implementd');
    }
    
    is_dirty() {
        const nodes = this.child_nodes().values();
        for (let node of nodes) {
            if (node.is_dirty()) {
                return true;
            }
        }
        return false;
    }
    
    clean() {
        this.child_nodes().map(node => node.clean());
    }
    
    static node_path_list_of(node_path) {
        if (typeof node_path === 'string') {
            return node_path.split('/');
        }
        return node_path;
    }
    
    node_at_path(node_path) {
        const node_path_list = Properties.node_path_list_of(node_path);
        let current_node = this;
        for (let node_name of node_path_list) {
            current_node = current_node.named_node(node_name);
            if (current_node === null) {
                console.log("node not found: " + node_path_list.join('/'));
                return null;
            }
            if (current_node.type !== PropNodeType.group) {
                return current_node;
            }
        }
        return current_node;
    }
    
    named_node(name) {
        if (this.children.has(name)) {
            return this.children.get(name);
        }
        return null;
    }
    
    property_at_path(node_path) {
        const node = this.node_at_path(node_path);
        if (node !== null) {
            if (node.type !== PropNodeType.group) {
                return node;
            } else {
                console.log('Node found, but not a property');
            }
        }
        return null;
    }
    
    group_at_path(node_path) {
        const node = this.node_at_path(node_path);
        if (node !== null) {
            if (node.type === PropNodeType.group) {
                return node;
            } else {
                console.log('Node found, but not a group');
            }
        }
        return null;
    }
    
    value_of(node_path) {
        const node = this.property_at_path(node_path);
        if (node !== null) {
            return node.value();
        } else {
            console.log('node not found');
        }
    }
    
    set_value_of(node_path, value, from_run_time = false) {
        const node = this.property_at_path(node_path);
        if (node !== null) {
            return node.set(value, undefined, from_run_time); //indexed issue
        }
    }
    
    as_description() {
        let d = super.as_description();
        d.set('children', Array.from(this.child_nodes().values()).map(node => node.as_description()));
        return d;
    }
    
    as_description_json() {
        return value_to_json(this.as_description());
    }
    
    static create_from_description_object(description_obj, parent=null) {
        let type = PropNodeType[description_obj.type];
        if (type === PropNodeType.group) {
            let p = new Properties(description_obj.name, parent);
            description_obj.children.forEach(child => {
                p.children.set(child.name, Properties.create_from_description_object(child, p));
            });
            return p;
        } else {
            let p = new Property(description_obj.name, type, parent, description_obj.default);
            p.range(description_obj.min, description_obj.max);
            p.names = description_obj.names; //not sure (is this part of a type?)
            return p;
        }
    }
    
    
    /*let props = new Properties('test');
       let ui = props.add_group('ui');
       ui.add("blur", PropNodeType.unsigned_int);
       ui.add("min_circle", PropNodeType.unsigned_int);
       let colours = ["blue", "green", "yellow", "orange", "pink"];
       let col = props.add_group('colours');
       colours.forEach(colour_name => {
           let sub = col.add_group(colour_name);
           sub.add('min_hsv', PropNodeType.hsv);
           sub.add('max_hsv', PropNodeType.hsv);
       });
       console.log(props.contents());
       console.log(props.contents_json());
       console.log(props.as_description());
       console.log(props.as_description_json());*/
    
    
}